plugins/plugin-java/che-plugin-java-plain/che-plugin-java-plain-server/src/main/java/org/eclipse/che/plugin/java/plain/server/inject/PlainJavaProjectSourceFolderWatcher.java

```diff
@@ -0,0 +1,197 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Reports the create/update/delete changes on project source folders to jdt.ls
+ *
+ * @author V. Rubezhny
+ */
+public class PlainJavaProjectSourceFolderWatcher {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(PlainJavaProjectSourceFolderWatcher.class);
+
+  private static final List<String> EMPTY_SOURCE_FOLDERS = emptyList();
+
+  private final FileWatcherManager manager;
+  private final FileWatcherByPathMatcher matcher;
+  private final EventService eventService;
+  private final ExecutorService executorService;
+
+  private final CopyOnWriteArrayList<Integer> watcherIds = new CopyOnWriteArrayList<>();
+  private final JavaLanguageServerExtensionService extensionService;
+  private final CopyOnWriteArrayList<LanguageServer> languageServers = new CopyOnWriteArrayList<>();
+
+  private final ProjectManager projectManager;
+
+  @Inject
+  public PlainJavaProjectSourceFolderWatcher(
+      FileWatcherManager manager,
+      FileWatcherByPathMatcher matcher,
+      EventService eventService,
+      JavaLanguageServerExtensionService extensionService,
+      ProjectManager projectManager) {
+    this.manager = manager;
+    this.matcher = matcher;
+    this.eventService = eventService;
+    this.extensionService = extensionService;
+    this.projectManager = projectManager;
+    this.executorService =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("WorkspaceUpdater-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
+                .build());
+  }
+
+  @PostConstruct
+  protected void startWatchers() {
+    int watcherId =
+        manager.registerByMatcher(
+            folderMatcher(),
+            s -> report(s, FileChangeType.Created),
+            s -> report(s, FileChangeType.Changed),
+            s -> report(s, FileChangeType.Deleted));
+
+    watcherIds.add(watcherId);
+    eventService.subscribe(this::onServerInitialized, LanguageServerInitializedEvent.class);
+    eventService.subscribe(this::onProjectUpdated, ProjectUpdatedEvent.class);
+  }
+
+  @PreDestroy
+  public void stopWatchers() {
+    watcherIds.stream().forEach(id -> manager.unRegisterByMatcher(id));
+  }
+
+  private void onServerInitialized(LanguageServerInitializedEvent event) {
+    LOG.debug(
+        "Registering source folder watching operations for language server : {}", event.getId());
+
+    String id = event.getId();
+    LanguageServer languageServer = event.getLanguageServer();
+    languageServers.add(languageServer);
+  }
+
+  private void onProjectUpdated(ProjectUpdatedEvent event) {
+    executorService.submit(
+        () -> {
+          setInitialPathsToWatch(prefixURI(event.getProjectPath()));
+        });
+  }
+
+  private PathMatcher folderMatcher() {
+    return it -> isSourceFolder(it);
```

Comment:
This is causing us to communicate with jdt.ls for every folder change in the file system. I think it would be more efficient to send every folder change to jdt.ls and let jdt.ls do the filtering.
Also, are we interested in every change or only in create/delete?